src/track/macro.cpp

```diff
@@ -0,0 +1,151 @@
+#include "track/macro.h"
+
+#include <QDebug>
+
+#include "util/assert.h"
+
+proto::Macro_Action* MacroAction::serialize() const {
+    auto serialized = new proto::Macro_Action();
+    serialized->set_position(position);
+    serialized->set_target(target);
+    serialized->set_type(static_cast<uint32_t>(type));
+    return serialized;
+}
+
+Macro::Macro(QList<MacroAction> actions, QString label, State state, int dbId)
+        : m_bDirty(false),
+          m_iId(dbId),
+          m_actions(actions),
+          m_label(label),
+          m_state(state) {
+}
+
+// static
+QList<MacroAction> Macro::deserialize(QByteArray serialized) {
+    proto::Macro macroProto = proto::Macro();
+    macroProto.ParseFromArray(serialized.data(), serialized.length());
+    QList<MacroAction> result;
+    result.reserve(macroProto.actions_size());
+    for (const proto::Macro_Action& action : macroProto.actions()) {
+        result.append(MacroAction(action.position(), action.target()));
+    }
+    return result;
+}
+
+// static
+QByteArray Macro::serialize(QList<MacroAction> actions) {
+    proto::Macro macroProto;
+    auto actionsProto = macroProto.mutable_actions();
+    for (const MacroAction& action : actions) {
+        actionsProto->AddAllocated(action.serialize());
+    }
+    auto string = macroProto.SerializeAsString();
+    return QByteArray(string.data(), string.length());
+}
+
+// static
+int Macro::getFreeSlot(QList<int> taken) {
+    int number = 1;
+    while (taken.contains(number))
+        number++;
+    return number;
+}
+
+bool Macro::isDirty() const {
+    return m_bDirty;
+}
+
+int Macro::getId() const {
+    return m_iId;
+}
+
+QString Macro::getLabel() const {
+    return m_label;
+}
+
+void Macro::setLabel(QString label) {
+    m_bDirty = true;
+    m_label = label;
+}
+
+bool Macro::isEnabled() const {
+    return m_state.testFlag(StateFlag::Enabled);
+}
+
+bool Macro::isLooped() const {
+    return m_state.testFlag(StateFlag::Looped);
+}
+
+const Macro::State& Macro::getState() const {
+    return m_state;
+}
+
+void Macro::setState(StateFlag flag, bool enable) {
+    if (m_state.testFlag(flag) != enable) {
+        m_bDirty = true;
+        m_state.setFlag(flag, enable);
+    }
+}
+
+bool Macro::isEmpty() const {
+    return m_actions.isEmpty();
+}
+
+unsigned int Macro::size() const {
+    return m_actions.size();
+}
+
+const QList<MacroAction>& Macro::getActions() const {
+    return m_actions;
+}
+
+void Macro::addAction(const MacroAction& action) {
+    m_bDirty = true;
+    m_actions.append(action);
+}
+
+double Macro::getStartSamplePos() const {
+    VERIFY_OR_DEBUG_ASSERT(!isEmpty()) {
+        return 0;
```

Comment:
I think in cuecontrol we use Cue::kNoPosition (= -1) for a non-existing position. Using 0 here would be inconsistent.

Reply:
This is not a valid state, only an emergency fallback. So I don't think that really matters.

Reply:
If it doesn't matter, then let's make this consistent and use -1.

Reply:
Negative sample positions may be problematic. Just for the unthinkable case that this ever happens, I want to default to 0, i.e. beginning of the track.
Bringing it in line with an unpositioned cue does not seem sensible to me, since this is not supposed to ever be negative.